parser/lexer.go

```diff
@@ -0,0 +1,373 @@
+package parser
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"unicode"
+)
+
+var _ = yyLexer(&Scanner{})
+
+// Pos represents the position of a token.
+type Pos struct {
+	Line   int
+	Col    int
+	Offset int
+}
+
+// Scanner implements the yyLexer interface
+type Scanner struct {
+	r   reader
+	buf bytes.Buffer
+
+	errs []error
+}
+
+// Errors returns the errors during a scan.
+func (s *Scanner) Errors() []error {
+	return s.errs
+}
+
+// reset resets the sql string to be scanned.
+// Scanner satisfies yyReset interface.
+func (s *Scanner) reset(sql string) yyLexerXX {
+	s.r = reader{s: sql}
+	return s
+}
+
+// Errorf tells scanner something is wrong.
+// Scanner satisfies yyLexer interface which need this function.
+func (s *Scanner) Errorf(format string, a ...interface{}) {
+	str := fmt.Sprintf(format, a...)
+	val := s.r.s[s.r.pos().Offset:]
+	err := fmt.Errorf("line %d column %d near \"%s\"%s", s.r.p.Line, s.r.p.Col, val, str)
+	s.errs = append(s.errs, err)
+}
+
+// Lex returns a token and store the token value in v.
+// Scanner satisfies yyLexer interface.
+func (s *Scanner) Lex(v *yySymType) int {
+	tok, pos, lit := s.scan()
+	v.offset = pos.Offset
+	if tok == identifier {
+		if tok1, ok := tokenMap[strings.ToUpper(lit)]; ok {
+			return tok1
+		}
+	}
+	v.item = lit
+	if tok == unicode.ReplacementChar && s.r.eof() {
+		return 0
+	}
+	return tok
+}
+
+// NewScanner returns a new scanner object.
+func NewScanner(s string) *Scanner {
+	return &Scanner{r: reader{s: s}}
+}
+
+func (s *Scanner) skipWhitespace() byte {
+	return s.r.incAsLongAs(isWhitespace)
+}
+
+func (s *Scanner) scan() (tok int, pos Pos, lit string) {
+	ch0 := s.r.peek()
+	if isWhitespace(ch0) {
+		ch0 = s.skipWhitespace()
+	}
+	pos = s.r.pos()
+
+	if isIdentFirstChar(ch0) {
+		return s.scanIdent()
+	} else if isDigit(ch0) || ch0 == '.' {
+		return s.scanNumber()
+	} else if ch0 == '\'' || ch0 == '"' {
+		return s.scanString()
+	}
+
+	for _, v := range tokenPrefixTable {
```

Comment:
Yes, I agree.
It's not best practice,  We could using a trie data struct to make a lookup table or using other method, but let's keep it simple currently. `tokenPrefixTable` is not big and I think we can bear it.

The new lexer's compatibility have higher priority, make it run first, then make it faster and make it elegant.